ludwig/schema/features/augmentation/utils.py

```diff
@@ -0,0 +1,156 @@
+
+    Returns: Initialized dataclass field that converts a list with params to an augmentation config.
+    """
+
+    class AugmentationContainerMarshmallowField(fields.Field):
+        """Custom marshmallow field that deserializes a list for a valid augmentation config from the
+        augmentation_registry and creates a corresponding JSON schema for external usage."""
+
+        def _deserialize(self, value, attr, data, **kwargs):
+            assert isinstance(value, list), "Augmentation config must be a list."
+
+            augmentation_list = []
+            for augmentation in value:
+                augmentation_op = augmentation[TYPE]
+                augmentation_cls = get_augmentation_cls(feature_type, augmentation_op)
+                pre = augmentation_cls()
+                try:
+                    augmentation_list.append(pre.Schema().load(augmentation))
+                except (TypeError, ValidationError) as error:
+                    raise ValidationError(
+                        f"Invalid augmentation params: {value}, see `{pre}` definition. Error: {error}"
+                    )
+            return augmentation_list
+
+        @staticmethod
+        def _jsonschema_type_mapping():
+            return get_augmentation_list_jsonschema(feature_type)
+
+    try:
+        assert isinstance(default, list), "Augmentation config must be a list."
+        load_augmentation_list = []
+        dump_augmentation_list = []
+        for augmentation in default:
+            augmentation_op = augmentation[TYPE]
+            augmentation_cls = get_augmentation_cls(feature_type, augmentation_op)
+            pre = augmentation_cls()
+            try:
+                load_augmentation_list.append(pre.Schema().load(augmentation))
+                dump_augmentation_list.append(pre.Schema().dump(augmentation))
+            except (TypeError, ValidationError) as error:
+                raise ValidationError(f"Invalid augmentation params: {default}, see `{pre}` definition. Error: {error}")
+        load_default = load_augmentation_list
+        dump_default = dump_augmentation_list
+
+        return field(
+            metadata={
+                "marshmallow_field": AugmentationContainerMarshmallowField(
+                    allow_none=False,
+                    dump_default=dump_default,
+                    load_default=load_default,
+                )
+            },
+            default_factory=lambda: load_default,
+        )
+    except Exception as e:
+        raise ValidationError(f"Unsupported augmentation type. See augmentation_registry. " f"Details: {e}")
+
+
+@DeveloperAPI
+def get_augmentation_list_jsonschema(feature_type: str):
+    """This function returns a JSON augmentation schema.
+
+    Returns: JSON Schema
+    """
+    augmentation_types = sorted(list(get_augmentation_config_registry()[feature_type].keys()))
+    schema = {
+        "type": "array",
+        "minItems": 0,
```

Comment:
@jimthompson5802 This is why `[]` does not throw a validation error. You should also add a check to the `_deserialize` that explicitly checks the length of the list is >= 1 to match.

Reply:
I added the check to `_deserialize` and still do not generate `ValidationError` with empty list.

If I set `minItems`  to 1 and failure test catches the `ValidationError` with empthy list but the other tests that should work now fail.

Should I have set `minItems` to 1?
 
If I set `minItems` back to 0, I believe I can still catch the empty list situation but the code will need to be in `ECD.get_augmentation_pipelines()` method.  This means the "validation" logic not contained completely in the schema validation code base.

Reply:
Setting `minItems` back to zero and adding a test in `ECD.get_augmentation_pipelines()` work.  

All augmentation pipeline tests now pass include the test to confirm failure when augmentation pipeline is an empty list.

Reply:
@jimthompson5802 If that is the case, something a little funky is still going on. You should set `minItems` to 1 because requirement is what should throw the validation error for the empty list. Can you trace where precisely in the code the error is thrown for that case?

> added the check to _deserialize and still do not generate ValidationError with empty list

That is fine. The checks in `_deserialize` are the last line of defense - the error for the empty list should be thrown before reaching it.

> but the code will need to be in ECD.get_augmentation_pipelines() method. This means the "validation" logic not contained completely in the schema validation code base.

That makes sense, though we should be able to avoid this.

Reply:
OK.  I'll set minItems back to zero and trace where the error is thrown.

Reply:
@ksbrar   this commit https://github.com/ludwig-ai/ludwig/pull/2925/commits/4039e55777f25a86ad98655d213b8eef4159ce1c takes care of the issue.  this code was first put there when I first tried modifying the schema validation and before you helped me work through the issues.   

With the removal of the code and setting `minItems: 1`, the validation test now correctly flags an empty list as an error.

Reply:
Wonderful!